README.md
# sanji-ethernet
Ethernet model.

## Usage

### Release a New Version
Both `debian/changelog` and `bundle.json` will be updated by new version.

```
make release [DIST=] [VER=]
```

arguments:
 - [DIST]     optional; default is unstable
 - [VER]      optional; the version will be increased automatically by default

### Build Debian Package
```
make deb [SANJI_VER=]
```

arguments:
 - [SANJI_VER]  optional; sanji's version, default to 1.0



debian/changelog
sanji-bundle-ethernet (0.9.8) unstable; urgency=low

  * Using customized script instead of default script with hook script for
    dhclient to prevent DNS and default gateway be updated.

 -- Aeluin Chen <aeluin.chen@moxa.com>  Thu, 04 Jun 2015 17:50:35 +0800

sanji-bundle-ethernet (0.9.0) unstable; urgency=low

  * Initial Release.

 -- Aeluin Chen <aeluin.chen@moxa.com>  Tue, 02 Dec 2014 17:02:46 +0800


tests/test_ethernet.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-


import os
import sys
import logging
import unittest
from mock import patch

from sanji.connection.mockup import Mockup
from sanji.message import Message

try:
    sys.path.append(os.path.dirname(os.path.realpath(__file__)) + '/../')
    from ethernet import Ethernet
except ImportError as e:
    print os.path.dirname(os.path.realpath(__file__)) + '/../'
    print sys.path
    print e
    print "Please check the python PATH for import test module. (%s)" \
        % __file__
    exit(1)

dirpath = os.path.dirname(os.path.realpath(__file__))


def mock_ip_ifaddresses(arg):
    if "eth0" == arg:
        return {"mac": "78:ac:c0:c1:a8:fe",
                "link": 1,
                "inet": [{
                    "broadcast": "192.168.31.255",
                    "ip": "192.168.31.36",
                    "netmask": "255.255.255.0",
                    "subnet": "192.168.31.0"}]}
    elif "eth1" == arg:
        return {"mac": "78:ac:c0:c1:a8:ff",
                "link": 0,
                "inet": [{
                    "broadcast": "192.168.31.255",
                    "ip": "192.168.31.37",
                    "netmask": "255.255.255.0",
                    "subnet": "192.168.31.0"}]}
    else:
        raise ValueError


class TestEthernetClass(unittest.TestCase):

    @patch("ethernet.ip.interfaces")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def setUp(self, mock_ifaddresses, mock_ifconfig, mock_ifupdown,
              mock_interfaces):
        # Setup the mock
        mock_interfaces.return_value = ["eth0", "eth1"]
        mock_ifaddresses.side_effect = mock_ip_ifaddresses

        self.name = "ethernet"
        self.bundle = Ethernet(connection=Mockup())

    def tearDown(self):
        self.bundle.stop()
        self.bundle = None
        try:
            os.remove("%s/data/%s.json" % (dirpath, self.name))
        except OSError:
            pass

        try:
            os.remove("%s/data/%s.json.backup" % (dirpath, self.name))
        except OSError:
            pass

    @patch("ethernet.ip.interfaces")
    def test__init__no_iface(self, mock_interfaces):
        """
        init: no interface
        """
        mock_interfaces.return_value = []

        with self.assertRaises(ValueError):
            self.bundle.init()

    @patch("ethernet.ip.interfaces")
    def test__init__no_conf(self, mock_interfaces):
        """
        init: no configuration file
        """
        mock_interfaces.return_value = ["eth0"]

        with self.assertRaises(IOError):
            with patch("ethernet.ModelInitiator") as mock_modelinit:
                mock_modelinit.side_effect = IOError
                self.bundle.init()

    @patch("ethernet.ip.ifaddresses")
    def test__load__current_conf(self, mock_ifaddresses):
        """
        load: current configuration file
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses

        ifaces = ["eth0", "eth1"]
        self.bundle.load(dirpath, ifaces)
        self.assertEqual(2, len(self.bundle.model.db))
        for iface in self.bundle.model.db:
            ifname = "eth%d" % (iface["id"]-1)
            self.assertTrue(ifname in ifaces)

    @patch("ethernet.ip.ifaddresses")
    def test__load__backup_conf(self, mock_ifaddresses):
        """
        load: backup configuration file
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses
        os.remove("%s/data/%s.json" % (dirpath, self.name))

        ifaces = ["eth0", "eth1"]
        self.bundle.load(dirpath, ifaces)
        self.assertEqual(2, len(self.bundle.model.db))

    def test__load__no_conf(self):
        """
        load: no configuration file
        """
        # case: cannot load any configuration
        with self.assertRaises(IOError):
            self.bundle.load("%s/mock" % dirpath, [])

    def test__save(self):
        """
        save: tested in init()
        """
        # Already tested in init()
        pass

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    def test__apply__iface_down(self, mock_ifconfig, mock_ifupdown):
        """
        apply: set the interface to "down"
        """
        # TODO: how to determine if setting success
        data = {
            "id": 1,
            "enable": 0
        }
        self.bundle.apply(data)

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    def test__apply__iface_up_static(self, mock_ifconfig, mock_ifupdown):
        """
        apply: set the interface to "up" with static IP
        """
        # TODO: how to determine if setting success
        data = {
            "id": 1,
            "enable": 1,
            "enableDhcp": 0,
            "ip": "192.168.31.39",
            "netmask": "255.255.255.0",
            "gateway": "192.168.31.254"
        }
        self.bundle.apply(data)

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    def test__apply__iface_up_dhcp(self, mock_ifconfig, mock_ifupdown):
        """
        apply: set the interface to "up" with dhcp enabled
        """
        # TODO: how to determine if setting success
        data = {
            "id": 1,
            "enable": 1,
            "enableDhcp": 1
        }
        self.bundle.apply(data)

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    def test__apply__iface_unknown(self, mock_ifconfig, mock_ifupdown):
        """
        apply: set an unknown interface
        """
        data = {
            "id": 0,
            "enable": 1
        }
        with self.assertRaises(ValueError):
            mock_ifupdown.side_effect = ValueError
            mock_ifconfig.side_effect = ValueError
            self.bundle.apply(data)

    @patch("ethernet.ip.ifaddresses")
    def test__read(self, mock_ifaddresses):
        """
        read: read current status and mac
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses

        data = self.bundle.read(1)
        self.assertEqual(1, data["currentStatus"])
        self.assertEqual("78:ac:c0:c1:a8:fe", data["mac"])

    @patch("ethernet.ip.ifaddresses")
    def test__read__unknown_iface(self, mock_ifaddresses):
        """
        read: no such interface
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses

        data = self.bundle.read(3)
        self.assertEqual(None, data)
        data = self.bundle.read(0)
        self.assertEqual(None, data)

    @patch("ethernet.ip.ifaddresses")
    def test__get__collection(self, mock_ifaddresses):
        """
        get (/network/ethernets): collection
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses
        message = Message({"data": {}, "query": {}, "param": {}})

        # case: collection
        def resp(code=200, data=None):
            self.assertEqual(200, code)
            self.assertEqual(2, len(data))
        self.bundle.get(message=message, response=resp, test=True)

    @patch("ethernet.ip.ifaddresses")
    def test__get__by_id(self, mock_ifaddresses):
        """
        get (/network/ethernets?id=2)
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses
        message = Message({"data": {}, "query": {}, "param": {}})

        # case: by id
        def resp(code=200, data=None):
            self.assertEqual(200, code)
            self.assertEqual(2, data["id"])
            self.assertEqual(0, data["currentStatus"])
            self.assertEqual("78:ac:c0:c1:a8:ff", data["mac"])
        message.query["id"] = 2
        self.bundle.get(message=message, response=resp, test=True)

    @patch("ethernet.ip.ifaddresses")
    def test__get_by_id(self, mock_ifaddresses):
        """
        get_by_id (/network/ethernets/1)
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses
        message = Message({"data": {}, "query": {}, "param": {}})

        # case: get an interface successfully
        def resp(code=200, data=None):
            self.assertEqual(200, code)
            self.assertEqual(1, data["id"])
            self.assertEqual(1, data["currentStatus"])
            self.assertEqual("78:ac:c0:c1:a8:fe", data["mac"])
        message.param["id"] = 1
        self.bundle.get_by_id(message=message, response=resp, test=True)

    @patch("ethernet.ip.ifaddresses")
    def test__get_by_id__unknown_iface(self, mock_ifaddresses):
        """
        get_by_id (/network/ethernets/3): unknown interface
        """
        mock_ifaddresses.side_effect = mock_ip_ifaddresses
        message = Message({"data": {}, "query": {}, "param": {}})

        # case: get an interface successfully
        def resp(code=200, data=None):
            self.assertEqual(404, code)
            self.assertEqual(data, {"message": "No such device."})
        message.param["id"] = 3
        self.bundle.get_by_id(message=message, response=resp, test=True)

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put__no_data(self, mock_ifaddresses, mock_ifconfig,
                           mock_ifupdown):
        """
        put (/network/ethernets): no data attribute
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        message = Message({"query": {}, "param": {}})

        # case: no data attribute
        def resp(code=200, data=None):
            self.assertEqual(400, code)
        self.bundle.put(message, response=resp, test=True)

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put__invalid_json(self, mock_ifaddresses, mock_ifconfig,
                                mock_ifupdown):
        """
        put (/network/ethernets): invalid json schema
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        # case: invalid json schema

        def resp(code=200, data=None):
            self.assertEqual(400, code)
        message = Message({"data": {}, "query": {}, "param": {}})
        self.bundle.put(message, response=resp, test=True)

        message = Message({"data": [], "query": {}, "param": {}})
        self.bundle.put(message, response=resp, test=True)

        message.data.append({"id": 0, "enable": 1})
        self.bundle.put(message, response=resp, test=True)

    @patch("ethernet.time.sleep")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put__partial_success(self, mock_ifaddresses, mock_ifconfig,
                                   mock_ifupdown, mock_sleep):
        """
        put (/network/ethernets): one interface is not exist
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        def mock_put(resource, data):
            pass
        self.bundle.publish.put = mock_put
        self.bundle.publish.event.put = mock_put

        # case: one interface is not exist (bulk); exist one will be updated
        message = Message({"data": [], "query": {}, "param": {}})

        # always true for response reply before apply the settings
        def resp(code=200, data=None):
            self.assertEqual(200, code)
        message.data.append({"id": 1, "enable": 1, "ip": u"192.168.31.36"})
        message.data.append({"id": 3, "enable": 1})
        self.bundle.put(message, response=resp, test=True)
        data = self.bundle.read(1, test=True)
        self.assertEqual("192.168.31.36", data["ip"])

    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put__unknown_ifaces(self, mock_ifaddresses, mock_ifconfig,
                                  mock_ifupdown):
        """
        put (/network/ethernets): all interfaces are not exist
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        message = Message({"data": [], "query": {}, "param": {}})

        def resp(code=200, data=None):
            self.assertEqual(400, code)
        message.data.append({"id": 0, "enable": 1})
        message.data.append({"id": 3, "enable": 1})
        self.bundle.put(message, response=resp, test=True)

    @patch("ethernet.time.sleep")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put(self, mock_ifaddresses, mock_ifconfig, mock_ifupdown,
                  mock_sleep):
        """
        put (/network/ethernets)
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        message = Message({"data": [], "query": {}, "param": {}})

        def mock_put(resource, data):
            pass
        self.bundle.publish.put = mock_put

        def resp(code=200, data=None):
            self.assertEqual(200, code)
            self.assertEqual(2, len(data))
        message.data.append({"id": 1, "enable": 1, "ip": u"192.168.31.37"})
        message.data.append({"id": 2, "enable": 1, "ip": u"192.168.41.37"})

        def mock_event_put(resource, data):
            pass
        self.bundle.publish.event.put = mock_event_put
        self.bundle.put(message, response=resp, test=True)

        data = self.bundle.read(1, test=True)
        self.assertEqual("192.168.31.37", data["ip"])
        data = self.bundle.read(2, test=True)
        self.assertEqual("192.168.41.37", data["ip"])

    @patch("ethernet.time.sleep")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put__by_id(self, mock_ifaddresses, mock_ifconfig, mock_ifupdown,
                         mock_sleep):
        """
        put (/network/ethernets): by id
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        message = Message({"data": [], "query": {}, "param": {}})

        def mock_put(resource, data):
            pass
        self.bundle.publish.put = mock_put

        def resp(code=200, data=None):
            self.assertEqual(200, code)
            self.assertEqual(1, data["enable"])
            self.assertEqual("192.168.31.39", data["ip"])
        message = Message({"data": {}, "query": {}, "param": {}})
        message.param["id"] = 1
        message.data["id"] = 1
        message.data["enable"] = 1
        message.data["ip"] = u"192.168.31.39"

        def mock_event_put(resource, data):
            pass
        self.bundle.publish.event.put = mock_event_put
        self.bundle.put(message, response=resp, test=True)

    @patch("ethernet.time.sleep")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put_by_id__invalid_json(self, mock_ifaddresses, mock_ifconfig,
                                      mock_ifupdown, mock_sleep):
        """
        put_by_id (/network/ethernets/1): invalid json schema
        "data": {
            "id": 1,
            ...
        }
        """
        message = Message({"query": {}, "param": {}})

        # no data attribute, other case already tested in "test_put()"
        def resp(code=200, data=None):
            self.assertEqual(400, code)
        self.bundle.put_by_id(message, response=resp, test=True)

    @patch("ethernet.time.sleep")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put_by_id__unknown_iface(self, mock_ifaddresses, mock_ifconfig,
                                       mock_ifupdown, mock_sleep):
        """
        put_by_id (/network/ethernets/3): unknown interface
        "data": {
            "id": 1,
            ...
        }
        """
        message = Message({"data": {}, "query": {}, "param": {}})

        def mock_put(resource, data):
            pass
        self.bundle.publish.put = mock_put

        def resp(code=200, data=None):
            self.assertEqual(404, code)
        message.data["id"] = 3
        message.data["enable"] = 0
        message.data["ip"] = u"192.168.31.37"
        self.bundle.put_by_id(message, response=resp, test=True)

    @patch("ethernet.time.sleep")
    @patch("ethernet.ip.ifupdown")
    @patch("ethernet.ip.ifconfig")
    @patch("ethernet.ip.ifaddresses")
    def test__put_by_id(self, mock_ifaddresses, mock_ifconfig, mock_ifupdown,
                        mock_sleep):
        """
        put_by_id (/network/ethernets/1)
        "data": {
            "id": 1,
            ...
        }
        """
        message = Message({"data": {}, "query": {}, "param": {}})

        def mock_put(resource, data):
            pass
        self.bundle.publish.put = mock_put

        def resp(code=200, data=None):
            self.assertEqual(404, code)
        message.data["id"] = 1
        message.data["enable"] = 0
        message.data["ip"] = u"192.168.31.40"
        self.bundle.put_by_id(message, response=resp, test=True)
        data = self.bundle.read(1, test=True)
        self.assertEqual("192.168.31.40", data["ip"])

    def test__put_dhcp_info__invalid_json(self):
        """
        put_dhcp_info (/network/interfaces/dhcp): invalid json schema
        "data": {
            "name": "",
            "ip": "",
            "netmask": "",
            "subnet": "",
            "dns": [],
            "gateway": ""
        }
        """
        message = Message({"query": {}, "param": {}})

        with self.assertRaises(ValueError):
            self.bundle.put_dhcp_info(message, test=True)

    def test__put_dhcp_info__unknown_iface(self):
        """
        put_dhcp_info (/network/interfaces/dhcp): unknown interface
        "data": {
            "name": "",
            "ip": "",
            "netmask": "",
            "subnet": "",
            "dns": [],
            "gateway": ""
        }
        """
        message = Message({"data": {}, "query": {}, "param": {}})
        message.data["name"] = "eth2"
        message.data["ip"] = "192.168.41.3"
        message.data["netmask"] = "255.255.255.0"
        message.data["gateway"] = "192.168.41.254"
        message.data["dns"] = ["8.8.8.8"]
        with self.assertRaises(ValueError):
            self.bundle.put_dhcp_info(message, test=True)

    @patch("ethernet.ip.ifaddresses")
    def test__put_dhcp_info(self, mock_ifaddresses):
        """
        put_dhcp_info (/network/interfaces/dhcp)
        "data": {
            "name": "",
            "ip": "",
            "netmask": "",
            "subnet": "",
            "dns": [],
            "gateway": ""
        }
        """
        message = Message({"data": {}, "query": {}, "param": {}})
        message.data["name"] = "eth1"
        message.data["ip"] = "192.168.41.3"
        message.data["netmask"] = "255.255.255.0"
        message.data["gateway"] = "192.168.41.254"
        message.data["dns"] = ["8.8.8.8"]
        self.bundle.put_dhcp_info(message, test=True)

        data = self.bundle.read(2, test=True)
        self.assertEqual("192.168.41.3", data["ip"])
        self.assertEqual("255.255.255.0", data["netmask"])
        self.assertEqual("192.168.41.254", data["gateway"])
        self.assertEqual(["8.8.8.8"], data["dns"])


if __name__ == "__main__":
    FORMAT = '%(asctime)s - %(levelname)s - %(lineno)s - %(message)s'
    logging.basicConfig(level=20, format=FORMAT)
    logger = logging.getLogger('Ethernet Test')
    unittest.main()


tools/ethernet-dhcp-updater.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
import sys
import subprocess
import getopt
import json
# from random import randint


DHCP_RES = "/network/ethernets/:iface/dhcp"
IFACE_RES = "/network/interfaces"

if __name__ == "__main__":
    try:
        opts, args = getopt.getopt(
            sys.argv[1:], "-i:",
            ["ip=", "netmask=", "subnet=", "gateway=", "dns="])
    except getopt.GetoptError:
        exit(-1)

    data = {
        "code": 200,
        "method": "put",
        "resource": DHCP_RES,
        "data": {
        }
    }

    for opt, arg in opts:
        if opt == "-i":
            data["data"]["name"] = arg
            id = int(arg.replace("eth", "")) + 1
            data["resource"] = DHCP_RES.replace(":iface", str(id))
        elif opt == "--ip":
            data["data"]["ip"] = arg
        elif opt == "--netmask":
            data["data"]["netmask"] = arg
        elif opt == "--subnet":
            data["data"]["subnet"] = arg
        elif opt == "--gateway":
            data["data"]["gateway"] = arg
        elif opt == "--dns":
            data["data"]["dns"] = arg.split()

    # send event to ethernet
    subprocess.Popen(
        ["mosquitto_pub",
         "-t", "/controller",
         "-m", "%s" % json.dumps(data, indent=2)],
        stdout=subprocess.PIPE)

    # send event to views
    data["resource"] = IFACE_RES
    subprocess.Popen(
        ["mosquitto_pub",
         "-t", "/controller",
         "-m", "%s" % json.dumps(data, indent=2)],
        stdout=subprocess.PIPE)


ethernet.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

import os
import copy
import time
import logging
from sanji.core import Sanji
from sanji.core import Route
from sanji.connection.mqtt import Mqtt
from sanji.model_initiator import ModelInitiator
from voluptuous import Schema
from voluptuous import Optional, Extra
# from voluptuous import Required
# from voluptuous import All
from voluptuous import In, Range, Any
# from voluptuous import Length, In, Range
import ip.addr as ip


_logger = logging.getLogger("sanji.ethernet")


def merge(dest, src, path=None):
    """Merges src into desc."""
    if path is None:
        path = []
    for key in src:
        if key in dest:
            if isinstance(dest[key], dict) and \
               isinstance(src[key], dict):
                merge(dest[key], src[key], path + [str(key)])
            elif dest[key] == src[key]:
                pass  # same leaf value
            else:
                dest[key] = src[key]
        else:
            dest[key] = src[key]
    return dest


class Ethernet(Sanji):
    """
    A model to handle Ethernet interfaces' configuration.

    Attributes:
        model: Ethernet interfaces' database with json format.
    """
    def init(self, *args, **kwargs):
        try:  # pragma: no cover
            bundle_env = kwargs["bundle_env"]
        except KeyError:
            bundle_env = os.getenv("BUNDLE_ENV", "debug")

        self.path_root = os.path.abspath(os.path.dirname(__file__))
        if bundle_env == "debug":  # pragma: no cover
            self.path_root = "%s/tests" % self.path_root

        # Find all ethernet interfaces and load the configuration
        ifaces = ip.interfaces()
        ifaces = [x for x in ifaces if x.startswith("eth")]
        if 0 == len(ifaces):
            _logger.info("No interfaces to be configured.")
            self.stop()
            raise ValueError("No interfaces to be configured.")

        try:
            self.load(self.path_root, ifaces)
        except:
            self.stop()
            raise IOError("Cannot load any configuration.")

        # Apply the configuration
        for iface in self.model.db:
            self.apply(iface)

    def run(self):
        for iface in self.model.db:
            self.publish.event.put("/network/interfaces", data=iface)

    def load(self, path, ifaces):
        """
        Load the configuration. If configuration is not installed yet,
        initialise them with the given interfaces and install them.

        Args:
            path: Path for the bundle, the configuration should be located
                under "data" directory.
            ifaces: A list of interfaces name.
        """
        self.model = ModelInitiator("ethernet", path, backup_interval=-1)
        if not self.model.db:
            raise IOError("Cannot load any configuration.")

        # Initialise the interfaces
        # TODO: 1. type is always "WAN" for CS
        #       2. 1st iface's type is "LAN" with dhcp server; another is "WAN"
        if 1 == len(self.model.db) and "id" not in self.model.db[0]:
            _logger.debug("factory install")
            default_db = self.model.db.pop()
            ip_3_def = int(default_db["ip"].split(".")[2]) - 1
            for iface in ifaces:
                ifaddr = ip.ifaddresses(iface)
                db = copy.deepcopy(default_db)
                db["name"] = iface
                db["id"] = int(iface.replace("eth", "")) + 1

                ip_3 = ip_3_def + db["id"]
                db["ip"] = "192.168.%d.127" % ip_3
                db["subnet"] = "192.168.%d.0" % ip_3
                db["gateway"] = "192.168.%d.254" % ip_3

                db["currentStatus"] = ifaddr["link"]
                db["mac"] = ifaddr["mac"]
                self.model.db.append(db)
            self.save()

    def save(self):
        """
        Save and backup the configuration.
        """
        self.model.save_db()
        self.model.backup_db()

    def apply(self, data):
        """
        Apply the configuration to an interface.

        Args:
            data: Information for the interface to be applied (with dictionary
                format)
        """
        iface = "eth%d" % (data["id"]-1)

        ip.ifupdown(iface, True if data["enable"] else False)
        if not data["enable"]:
            return

        if data["enableDhcp"]:
            ip.ifconfig(iface, True, script="%s/hooks/dhclient-script" %
                        self.path_root)
        else:
            ip.ifconfig(iface, False, data["ip"], data["netmask"],
                        data["gateway"])

    def read(self, id, test=False):
        """
        Read the setting for an interface.

        Args:
            id: Interface id, interface name will be eth(id+1).
        """
        for data in self.model.db:
            if data["id"] == id:
                break
        else:
            return None

        # deepcopy to prevent settings be modified
        data = copy.deepcopy(data)

        iface = "eth%d" % (data["id"]-1)
        ifaddr = ip.ifaddresses(iface)
        data["currentStatus"] = ifaddr["link"]
        data["mac"] = ifaddr["mac"]

        # """Use configuration data instead of realtime retrieving
        if True == test:
            return data

        data["ip"] = ""
        data["netmask"] = ""
        data["subnet"] = ""
        data["broadcast"] = ""
        # data["gateway"] = ""
        if ifaddr["inet"] and len(ifaddr["inet"]):
            data["ip"] = ifaddr["inet"][0]["ip"]
            data["netmask"] = ifaddr["inet"][0]["netmask"]
            if "subnet" in ifaddr["inet"][0]:
                data["subnet"] = ifaddr["inet"][0]["subnet"]
            else:
                data.pop("subnet")
            if "broadcast" in ifaddr["inet"][0]:
                data["broadcast"] = ifaddr["inet"][0]["broadcast"]
            elif "broadcast" in data:
                data.pop("broadcast")
        # """
        return data

    @staticmethod
    def schema_validate(message):
        """
        Validate the received data, ensure the schema is correct.
        """
        # TODO: ip validation
        schema = Schema({
            "id": Range(min=1),
            "enable": In(frozenset([0, 1])),
            Optional("type"): In(frozenset([0, 1])),
            Optional("enableDhcp"): In(frozenset([0, 1])),
            Optional("ip"): Any(str, unicode),
            Optional("netmask"): Any(str, unicode),
            Optional("subnet"): Any(str, unicode),
            Optional("gateway"): Any(str, unicode),
            Optional("dns"): [Any(str, unicode)],
            Extra: object
        }, required=True)

        if not hasattr(message, "data"):
            raise KeyError("Invalid input: \"data\" attribute is required.")

        if type(message.data) is list:
            if 0 == len(message.data):
                raise KeyError("Invalid input: empty \"data\".")
            for item in message.data:
                try:
                    schema(item)
                except Exception, e:
                    raise KeyError("Invalid input: %s." % e)

        if type(message.data) is dict:
            try:
                schema(message.data)
            except Exception, e:
                raise KeyError("Invalid input: %s." % e)

    def _get_by_id(self, message, response):
        """
        /network/ethernets
        with id=#
        """
        ifinfo = self.read(int(message.param["id"]))
        if ifinfo:
            return response(data=ifinfo)
        return response(code=404, data={"message": "No such device."})

    @Route(methods="get", resource="/network/ethernets")
    def get(self, message, response):
        """
        collection: /network/ethernets
        id: /network/ethernets?id=#
        """

        # TODO: multiple id supported?
        if "id" in message.query:
            message.param["id"] = message.query["id"]
            return self._get_by_id(message=message, response=response)

        collection = []
        for iface in self.model.db:
            data = self.read(iface["id"])
            if data:
                collection.append(data)
        collection = sorted(collection, key=lambda k: k["id"])
        return response(data=collection)

        '''capability function removed
        capability = []
        for iface in self.model.db:
            capability.append(iface["id"])
        return response(data=capability)
        '''

    @Route(methods="get", resource="/network/ethernets/:id")
    def get_by_id(self, message, response):
        """
        /network/ethernets/1
        """
        return self._get_by_id(message=message, response=response)

    def merge_info(self, iface):
        """
        Merge the given interface information into database.
        """
        for item in self.model.db:
            if item["id"] == iface["id"]:
                return merge(item, iface)
        else:
            raise ValueError("No such device.")

    def _put_by_id(self, message, response):
        """
        /network/ethernets/1
        "data": {
            "id": 1,
            ...
        }
        """
        # TODO: status code should be added into error message
        # 1. no "data"
        # 2. data is dict() and no "enable"
        try:
            self.schema_validate(message)
        except Exception, e:
            _logger.debug(e, exc_info=True)
            return response(code=400,
                            data={"message": e.message})

        try:
            info = self.merge_info(message.data)
            response(data=info)

            time.sleep(2)
            self.publish.put("/system/remote", data={"enable": 0})

            self.apply(info)
            self.save()
            self.publish.event.put("/network/interfaces", data=info)

            time.sleep(2)
            self.publish.put("/system/remote", data={"enable": 1})
            # time.sleep(2)
            # return response(data=info)
        except Exception, e:
            self.publish.put("/system/remote", data={"enable": 1})
            time.sleep(2)
            return response(code=404, data={"message": e.message})

    @Route(methods="put", resource="/network/ethernets")
    def put(self, message, response):
        """
        bulk put: /network/ethernets
        "data": [
            {
                "id": 1,
                ...
            },
            {
                "id": 2,
                ...
            }
        ]
        """
        # TODO: status code should be added into error message
        # 1. no "data"
        # 2. data is dict() and no "enable"
        # 3. data is list() and len(data) > 0 and each element has no "enable"
        try:
            self.schema_validate(message)
        except Exception, e:
            _logger.debug(e, exc_info=True)
            return response(code=400,
                            data={"message": e.message})

        if "id" in message.param:
            return self._put_by_id(message=message, response=response)

        response(data=message.data)
        time.sleep(2)
        self.publish.put("/system/remote", data={"enable": 0})
        # error = None
        for iface in message.data:
            try:
                info = self.merge_info(iface)
                self.apply(info)
                self.model.save_db()
                self.publish.event.put("/network/interfaces", data=info)
            except Exception, e:
                # error = e.message
                pass
        self.model.backup_db()
        time.sleep(2)
        self.publish.put("/system/remote", data={"enable": 1})
        '''
        time.sleep(2)
        if error:
            return response(code=400, data={"message": error})
        return response(data=self.model.db)
        '''

    @Route(methods="put", resource="/network/ethernets/:id")
    def put_by_id(self, message, response):
        """
        /network/ethernets/1
        "data": {
            "id": 1,
            ...
        }
        """
        return self._put_by_id(message=message, response=response)

    put_dhcp_schema = Schema({
        "name": Any(str, unicode),
        "ip": Any(str, unicode),
        "netmask": Any(str, unicode),
        Optional("subnet"): Any(str, unicode),
        "gateway": Any(str, unicode),
        "dns": [Any(str, unicode)],
        Extra: object
    }, required=True)

    @Route(methods="put", resource="/network/interfaces/dhcp",
           schema=put_dhcp_schema)
    def put_dhcp_info(self, message):
        """
        /network/interfaces/dhcp
        "data": {
            "name": "",
            "ip": "",
            "netmask": "",
            "subnet": "",
            "dns": [],
            "gateway": ""
        }
        """
        # TODO: should be removed when schema worked for unittest
        if not hasattr(message, "data"):
            raise ValueError("Invalid input.")

        if "name" not in message.data or "eth" not in message.data["name"]:
            return

        message.data["id"] = int(message.data["name"].replace("eth", "")) + 1
        try:
            self.merge_info(message.data)
            _logger.debug(self.model.db)
            self.model.save_db()
        except Exception, e:
            raise ValueError("Invalid input: %s.", str(e))


if __name__ == "__main__":
    FORMAT = "%(asctime)s - %(levelname)s - %(lineno)s - %(message)s"
    logging.basicConfig(level=0, format=FORMAT)
    _logger = logging.getLogger("sanji.ethernet")

    ethernet = Ethernet(connection=Mqtt())
    ethernet.start()
